phantazm-neuron/src/main/java/com/github/phantazmnetwork/neuron/world/SpatialCollider.java

```diff
@@ -0,0 +1,236 @@
+
+                            //fast exit: we must have found the best possible in the entire volume
+                            if(fastExit.test(dY)) {
+                                return value;
+                            }
+
+                            overlapping.setPointer(variables.getFirstOrigin(), variables.getSecondOrigin(), nextY);
+                        }
+
+                        best = value;
+                    }
+                }
+            }
+            while (overlapping.hasNext());
+        }
+
+        return best;
+    }
+
+    private static boolean checkSolid(SolidPipe overlapping, Solid component, double centerX, double centerY,
+                                      double centerZ, double adjustedXZ, double adjustedXY, double adjustedYZ,
+                                      double dX, double dY, double dZ) {
+        Vec3F componentMin = component.getMin();
+        Vec3F componentMax = component.getMax();
+
+        double minX = overlapping.getFirst() + componentMin.getX() - centerX;
+        double minY = overlapping.getThird() + componentMin.getY() - centerY;
+        double minZ = overlapping.getSecond() + componentMin.getZ() - centerZ;
+
+        double maxX = overlapping.getFirst() + componentMax.getX() - centerX;
+        double maxY = overlapping.getThird() + componentMax.getY() - centerY;
+        double maxZ = overlapping.getSecond() + componentMax.getZ() - centerZ;
+
+        return checkAxis(adjustedXZ, dX, dZ, minX, minZ, maxX, maxZ) && checkAxis(adjustedXY, dX, dY, minX, minY, maxX,
+                maxY) && checkAxis(adjustedYZ, dZ, dY, minZ, minY, maxZ, maxY);
+    }
+
+    @Override
+    public double highestCollisionAlong(double x, double y, double z, double width, double height, double depth,
+                                        double dX, double dY, double dZ) {
+        return collisionCheck(x, y, z, width, height, depth, dX, dY, dZ, Double.NEGATIVE_INFINITY, (solid, val) -> val +
+                solid.getMax().getY(), (value, best) -> value > best, solid -> solid.getMax().getY() == 1.0F,
+                val -> val < 0);
+    }
+
+    @Override
+    public double lowestCollisionAlong(double x, double y, double z, double width, double height, double depth,
+                                       double dX, double dY, double dZ) {
+        return collisionCheck(x, y, z, width, height, depth, dX, dY, dZ, Double.POSITIVE_INFINITY, (solid, val) -> val +
+                solid.getMin().getY(), (value, best) -> value < best, solid -> solid.getMin().getY() == 0.0F,
+                val -> val > 0);
+    }
+
+    @Override
+    public double heightAt(int x, int y, int z) {
+        Solid solid = space.solidAt(x, y, z);
+        return solid == null ? y : y + solid.getMax().getY();
+    }
+
+    private static boolean checkAxis(double size, double dA, double dB, double minA, double minB, double maxA,
+                                     double maxB) {
+        if(dA == 0 && dB == 0) {
```

Comment:
what are the dA and dB

Reply:
direction vector for a particular plane to check along (like XY, YX, ZY, YZ, ...)
basically a projection of the actual direction vector (dX, dY, dZ) but the components are swapped depending on the quadrant